core/commands/pin.go

```diff
@@ -282,3 +272,76 @@ type RefKeyObject struct {
 type RefKeyList struct {
 	Keys map[string]RefKeyObject
 }
+
+func pinLsKeys(args []string, typeStr string, ctx context.Context, n *core.IpfsNode) (map[string]RefKeyObject, error) {
+
+	keys := make(map[string]RefKeyObject)
+
+	for _, p := range args {
+		dagNode, err := core.Resolve(ctx, n, path.Path(p))
+		if err != nil {
+			return nil, err
+		}
+
+		k, err := dagNode.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		pinType, pinned, err := n.Pinning.IsPinnedWithType(k, typeStr)
+		if err != nil {
+			return nil, err
+		}
+
+		if !pinned {
+			return nil, fmt.Errorf("Path '%s' is not pinned", p)
+		}
+
+		switch pinType {
+		case "direct", "indirect", "recursive", "internal":
+		default:
+			pinType = "indirect through " + pinType
```

Comment:
To me, it looks like `pinLsKeys` = `ipfs resolve | isPinnedWithType`.
I wonder why is this not in `IsPinnedWithType`?

`gc.GC()` uses a kind of `ipfs pin ls <hash>`, so there could be a refactor here.